### PR TITLE
Add accessors for container CPU shares

### PR DESCRIPTION
**Pull request by carterkozak**
This is difficult to test in a reasonable way.

### CODE REVIEW

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/CpuSharesAccessor.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
```
Comment (carterkozak) [inline jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/CpuSharesAccessor.java]:
```suggestion
 * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
```

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/CpuSharesAccessor.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.OptionalLong;
+
+public interface CpuSharesAccessor {
+
+    /**
+     * Returns the relative-weighted CPU shares detected by the JVM. In k8s, this should be a millicore value.
+     * Returns {@link OptionalLong#empty()} if information is unavailable (and this accessor is not returned itself
+     * if cpu shares are not supported).
+     */
+    OptionalLong getCpuShares();
```
Comment (schlosna) [inline jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/CpuSharesAccessor.java:28]: Might be worth adding to comments that https://bugs.openjdk.org/browse/JDK-8281181 / https://bugs.openjdk.org/browse/JDK-8281571 / https://bugs.openjdk.org/browse/JDK-8279484 /  https://github.com/openjdk/jdk/commit/e07fd395bdc314867886a621ec76cf74a5f76b89 immediately deprecates `UseContainerCpuShares` in JDK 19, 11.0.17, and 17.0.5

Comment (schlosna) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/CpuSharesAccessor.java:28]: May be also worth mentioning https://danluu.com/cgroup-throttling/

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java [hunk 3]
```diff
@@ -173,4 +188,22 @@ public long getCpuTimeNanoseconds(long threadId) {
             return threadManagementBean.getThreadCpuTime(threadId);
         }
     }
+
+    private static final class HotspotCpuSharesAccessor implements CpuSharesAccessor {
+
+        private final jdk.internal.platform.Metrics metrics = jdk.internal.platform.Metrics.systemMetrics();
+
+        boolean isEnabled() {
+            return metrics.getCpuShares() != -2;
```
Comment (schlosna) [inline jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:200]: From https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/platform/Metrics.java#L153-L169 I think we might want to adjust this to:

> shares value, -1 if the metric is not available or
>     *         -2 if cpu shares are not supported.

```suggestion
            return metrics.getCpuShares() >= 0;
```


- https://github.com/openjdk/jdk/blob/master/src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1Subsystem.java#L245-L250
- https://github.com/openjdk/jdk/blob/master/src/java.base/linux/classes/jdk/internal/platform/cgroupv2/CgroupV2Subsystem.java#L155-L180

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java [hunk 4]
```diff
@@ -104,6 +105,20 @@ public static Optional<ThreadUserTimeAccessor> threadUserTime() {
         }
     }
 
+    /**
+     * Returns an {@link HotspotCpuSharesAccessor}. This functionality is not supported on all java runtimes,
+     * and an {@link Optional#empty()} is returned in cases cpu share information is not supported.
+     */
+    public static Optional<CpuSharesAccessor> cpuShares() {
```
Comment (schlosna) [inline jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:115]: can we test this?

Comment (carterkozak) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:115]: I don't have a great way to do so with the tools we have in this repo, unless we match the environment circle uses -- unfortunately tests won't work locally in that case

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java (hunk 3 above)
Comment (carterkozak) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:200]: I wanted to distinguish between "I know what cpu shares are, but I'm not using them" from "I'm not sure how to tell if cpu shares are in play". In the former case, we return an accessor that provides OptionalLong.emty, and in the latter case we return an empty Optional.

Comment (carterkozak) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:200]: Would you prefer to avoid that distinction?

Comment (schlosna) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:200]: I assume you'll expose this as a gauge metric? Would you consume it as something like this?

```
JvmDiagnostics.cpuShares().ifPresent(shares -> registerGauge(() -> shares.getAsLong()))
```

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java (hunk 4 above)
Comment (schlosna) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:115]: I'm curious what circle returns on say OpenJDK 17.0.5 vs 17.0.4. We could add some JUnit assumptions around JDK version, OS, and compare to Runtime number of processors.

Comment (carterkozak) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:115]: I think this value will return the same results, but processor count may not (certainly not in our metrics)

File: jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java (hunk 3 above)
Comment (carterkozak) [reply jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/JvmDiagnostics.java:200]: Both a gauge metric, and our startup logging (the latter is where we can distinguish between unavailable and unsupported)

**Review by schlosna — APPROVED** [review]

**Pull request merged**